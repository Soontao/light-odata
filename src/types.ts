/* eslint-disable max-len */
import { RequestInit } from "node-fetch";
import { BatchRequest, ParsedResponse } from "./batch";
import { TokenRetrieveType } from "./oauth";
import { SystemQueryOptions } from "./params";
import { FormatODataDateTimedate } from "./util";

export type ODataVersion = "v2" | "v4";

export type HTTPMethod = "GET" | "HEAD" | "POST" | "PUT" | "DELETE" | "PATCH";

export type ODataVariant = "default" | "c4c" | "byd" | "cap" | "cpi" | "@odata/server" | "sap-gateway"

export const SAPNetweaverOData: Readonly<Array<ODataVariant>> = ["c4c", "byd", "sap-gateway"] as const;

export type FetchProxy = (url: string, init: RequestInit) => Promise<{
  /**
   * parsed body content
   */
  content: any;
  /**
   * original response object
   */
  response: Response;
}>;

export interface ODataNewOptionsOld {
  /**
   * metadata url
   *
   * @deprecated please try to use serviceEndpoint
   * @example `https://odata.org/V2/Northwind/Northwind.svc/$metadata`
   */
  metadataUri: string;
}

export interface ODataNewOptionsNew {
  /**
   *
   * odata service endpoint
   *
   * @example `https://odata.org/V2/Northwind/Northwind.svc/`
   */
  serviceEndpoint: string;
}

export interface ODataNewOptionsCommon {

  /**
   * credential configuration, basic/oauth (client credentials)
   */
  credential?: Credential;

  /**
   * fetch proxy of all request
   */
  fetchProxy?: FetchProxy;

  /**
   * auto process csrf token of c4c
   */
  processCsrfToken?: boolean;

  /**
   * common headers will be sent for each requests
   */
  commonHeaders?: { [headerName: string]: string };

  /**
   * the header key of csrf token
   *
   * @default 'x-csrf-token'
   */
  csrfTokenName?: string;

  /**
   * service variant, some behaviors will be changed by this option
   */
  variant?: ODataVariant;

  /**
   * odata version
   *
   * @default 'v2'
   */
  version?: ODataVersion;
}

export type ODataNewOptions = (ODataNewOptionsNew | ODataNewOptionsOld) & ODataNewOptionsCommon

export type ODataV4NewOptions = ODataNewOptions & {
  version: "v4"
}

export interface BatchRequestOptions<T> {
  /**
   * Collection Name
   */
  collection: string;
  /**
   * OData Entity ObjectID
   */
  id?: any;
  /**
   * OData system query options
   */
  params?: SystemQueryOptions;
  method?: HTTPMethod;
  /**
   * OData Entity Object
   */
  entity?: Partial<T>;
  /**
   * SAP OData need Content-Length but standard reject it
   */
  withContentLength?: boolean;
}

export interface ODataRequest {
  /**
   * the `entity set` name of data service
   */
  collection?: string,
  /**
   * GET for QUERY/READ; for QUERY, you can use params to control response data
   * PATCH for UPDATE
   * PUT for overrite
   * POST for CREATE
   * DELETE for delete
   */
  method?: HTTPMethod,
  params?: SystemQueryOptions,
}

export type ODataKeyPredicate = string | boolean | number | null | ODataValueObject | {
  [propertyKey: string]: ODataKeyPredicate
};

export interface ODataReadIDRequest extends ODataRequest {
  id: ODataKeyPredicate, /** object key in READ/UPDATE/DELETE */
  params?: SystemQueryOptions,
}

export interface ODataQueryRequest extends ODataRequest {
  params?: SystemQueryOptions, /** params in QUERY */
}

export interface ODataWriteRequest<T> extends ODataRequest {
  entity?: Partial<T> /** data object in CREATE/UPDATE */
  id?: ODataKeyPredicate;
  method: HTTPMethod
}

export interface ODataActionRequest extends ODataRequest {
  actionName: string;
  payload?: any;
  id: ODataKeyPredicate;
}

export interface ODataFunctionRequest extends ODataRequest {
  functionName: string;
  payload?: any;
  id: ODataKeyPredicate;
}

export interface ODataActionImportRequest extends ODataRequest {
  actionName: string;
  parameters?: { [key: string]: any };
}

export interface ODataFunctionImportRequest extends ODataRequest {
  functionName: string;
  parameters?: { [key: string]: any };
}


export interface PlainODataResponse {

  error?: { /** if error occured, node error will have value */
    code: string;
    message: {
      lang: string,
      value: string /** server error message */
    }
  }
  d?: any;

}

export type BatchPlainODataResponse<E = any> = PlainODataResponse & {
  d?: {
    __count?: string; /** $inlinecount values */
    results?: Array<E>;
  } & E;
}

export type PlainODataSingleResponse<E = any> = {

  /**
   * @version 2.0.0
   */
  d?: {
    __count?: string; /** $inlinecount values */
  } & E;

} & PlainODataResponse


export interface PlainODataMultiResponse<E = any> extends PlainODataResponse {

  d?: {
    __count?: string; /** $inlinecount values */
    results: Array<E>; /** result list/object */
  };

}

export interface Credential {
  /**
   * basic auth user name
   */
  username?: string;
  /**
   * basic password
   */
  password?: string;
  /**
   * oauth client id
   */
  clientId?: string;
  /**
   * oauth client secret
   */
  clientSecret?: string;
  /**
   * oauth token url
   */
  tokenUrl?: string;
  /**
   * oauth token retrieve type
   */
  tokenRetrieveType?: TokenRetrieveType;

  /**
   * traget scope
   */
  scope?: string;
}

/**
 * Same as Partial<T> but goes deeper and makes Partial<T> all its properties and sub-properties.
 */
export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Array<infer U> ? Array<DeepPartial<U>> : T[P] extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>> : DeepPartial<T[P]>;
};

/**
 * unwrap Promise Type
 */
export type UnwrapPromise<T> = T extends PromiseLike<infer U> ? U : T

/**
 * unwrap batch request
 */
export type UnwrapBatchRequest<T> = T extends BatchRequest<infer U> ? U : T


/**
 * unwrap batch request
 */
export type UnwrapParsedResponse<T> = T extends ParsedResponse<infer U> ? U : T


export type BatchRequests = Array<Promise<BatchRequest>>

export type BatchResponses<T> = Promise<{ [K in keyof T]: ParsedResponse<UnwrapBatchRequest<UnwrapPromise<T[K]>>> }>

/**
 * Edm value
 */
export abstract class ODataValueObject<T = any> {

  protected rawValue: T;

  protected constructor(rawValue: T) {
    this.rawValue = rawValue;
  }

  /**
   * to OData Uri string
   */
  abstract toString(): string;

  /**
   * to OData Json format value string
   *
   * @returns
   */
  public toJSONString(): string {
    return this.toString();
  }

}

export const ODataValueJSONReplacer = (key, value) => {
  if (value instanceof ODataValueObject) {
    return value.toJSONString();
  }
  return value;
};

/**
 * RawString
 */
export class RawString extends ODataValueObject<string> {

  public static from(str: string) {
    return new RawString(str);
  }

  toString(): string {
    return this.rawValue;
  }

}

export const RawValue = RawString;

export abstract class ODataDateBase extends ODataValueObject {

  protected _uriEncoded: boolean;

  protected constructor(date: Date, uriEncoded = true) {
    if (date instanceof Date) {
      super(date);
    } else {
      super(new Date(date));
    }
    this._uriEncoded = uriEncoded;
  }

  public toJSONString(): string {
    return FormatODataDateTimedate(this.rawValue);

  }
}

/**
 * Edm.DateTime
 */
export class ODataDateTime extends ODataDateBase {

  /**
   *
   * @param date date object
   * @param uriEncoded encode date string with `encodeURIComponent`, default is `true`
   * @returns
   */
  public static from(date: Date, uriEncoded?: boolean): ODataDateTime {
    return new ODataDateTime(date, uriEncoded);
  }

  public toString(): string {
    let inner: string = this.rawValue.toISOString().substring(0, 19);
    if (this._uriEncoded) {
      inner = encodeURIComponent(inner);
    }
    return `datetime'${inner}'`;
  }


}

/**
 * Edm.DateTimeOffset
 */
export class ODataDateTimeOffset extends ODataDateBase {

  /**
   *
   * @param date date object
   * @param uriEncoded encode date string with `encodeURIComponent`, use it if you requires
   * @returns
   */
  public static from(date: Date, uriEncoded?: boolean): ODataDateTimeOffset {
    return new ODataDateTimeOffset(date, uriEncoded);
  }

  public toString(): string {
    let inner: string = this.rawValue.toISOString();
    if (this._uriEncoded) {
      inner = encodeURIComponent(inner);
    }
    return `datetimeoffset'${inner}'`;
  }

}

export class Guid extends RawValue { }

export class EdmString extends ODataValueObject<string> {

  public static from(str: string): EdmString {
    return new EdmString(str);
  }

  public toString(): string {
    return `'${this.rawValue}'`;
  }

  public toJSONString(): string {
    return this.rawValue;
  }

}

export const EdmV2 = {
  DateTime: ODataDateTime,
  DateTimeOffset: ODataDateTimeOffset,
  RawString,
  Guid,
  String: EdmString,
};
