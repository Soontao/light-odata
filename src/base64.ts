/* eslint-disable max-len */
/* eslint-disable prefer-const */
// Array of bytes to Base64 string decoding
function b64ToUint6(nChr) {
  return nChr > 64 && nChr < 91 ?
    nChr - 65
    : nChr > 96 && nChr < 123 ?
      nChr - 71
      : nChr > 47 && nChr < 58 ?
        nChr + 4
        : nChr === 43 ?
          62
          : nChr === 47 ?
            63
            :
            0;
}

function base64DecToArr(sBase64: string, nBlocksSize?: number) {
  const sB64Enc = sBase64.replace(/[^A-Za-z0-9+/]/g, "");
  const nInLen = sB64Enc.length;
  const nOutLen = nBlocksSize ? Math.ceil((nInLen * 3 + 1 >> 2) / nBlocksSize) * nBlocksSize : nInLen * 3 + 1 >> 2;
  const taBytes = new Uint8Array(nOutLen);

  let nMod3;
  let nMod4;
  let nUint24 = 0;
  let nOutIdx = 0;
  for (let nInIdx = 0; nInIdx < nInLen; nInIdx++) {
    nMod4 = nInIdx & 3;
    nUint24 |= b64ToUint6(sB64Enc.charCodeAt(nInIdx)) << 6 * (3 - nMod4);
    if (nMod4 === 3 || nInLen - nInIdx === 1) {
      nMod3 = 0;
      while (nMod3 < 3 && nOutIdx < nOutLen) {
        taBytes[nOutIdx] = nUint24 >>> (16 >>> nMod3 & 24) & 255;
        nMod3++;
        nOutIdx++;
      }
      nUint24 = 0;

    }
  }

  return taBytes;
}

/* Base64 string to array encoding */
function uint6ToB64(nUint6: number) {
  return nUint6 < 26 ?
    nUint6 + 65
    : nUint6 < 52 ?
      nUint6 + 71
      : nUint6 < 62 ?
        nUint6 - 4
        : nUint6 === 62 ?
          43
          : nUint6 === 63 ?
            47
            :
            65;
}

function base64EncArr(aBytes) {
  let nMod3 = 2;
  let sB64Enc = "";

  const nLen = aBytes.length;
  let nUint24 = 0;
  for (let nIdx = 0; nIdx < nLen; nIdx++) {
    nMod3 = nIdx % 3;
    if (nIdx > 0 && (nIdx * 4 / 3) % 76 === 0) {
      sB64Enc += "\r\n";
    }

    nUint24 |= aBytes[nIdx] << (16 >>> nMod3 & 24);
    if (nMod3 === 2 || aBytes.length - nIdx === 1) {
      sB64Enc += String.fromCodePoint(uint6ToB64(nUint24 >>> 18 & 63), uint6ToB64(nUint24 >>> 12 & 63), uint6ToB64(nUint24 >>> 6 & 63), uint6ToB64(nUint24 & 63));
      nUint24 = 0;
    }
  }
  return sB64Enc.substring(0, sB64Enc.length - 2 + nMod3) + (nMod3 === 2 ? "" : nMod3 === 1 ? "=" : "==");
}

/* UTF-8 array to JS string and vice versa */

function utf8ArrToStr(aBytes) {
  let sView = "";
  let nPart;
  const nLen = aBytes.length;
  for (let nIdx = 0; nIdx < nLen; nIdx++) {
    nPart = aBytes[nIdx];
    sView += String.fromCodePoint(
      nPart > 251 && nPart < 254 && nIdx + 5 < nLen ? /* six bytes */
        /* (nPart - 252 << 30) may be not so safe in ECMAScript! So…: */
        (nPart - 252) * 1073741824 + (aBytes[++nIdx] - 128 << 24) + (aBytes[++nIdx] - 128 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
        : nPart > 247 && nPart < 252 && nIdx + 4 < nLen ? /* five bytes */
          (nPart - 248 << 24) + (aBytes[++nIdx] - 128 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
          : nPart > 239 && nPart < 248 && nIdx + 3 < nLen ? /* four bytes */
            (nPart - 240 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
            : nPart > 223 && nPart < 240 && nIdx + 2 < nLen ? /* three bytes */
              (nPart - 224 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
              : nPart > 191 && nPart < 224 && nIdx + 1 < nLen ? /* two bytes */
                (nPart - 192 << 6) + aBytes[++nIdx] - 128
                : /* nPart < 127 ? */ /* one byte */
                nPart
    );
  }
  return sView;
}

function strToUTF8Arr(sDOMStr: string): Uint8Array {
  let aBytes;
  let nChr;
  const nStrLen = sDOMStr.length;
  let nArrLen = 0;

  /* mapping… */
  for (let nMapIdx = 0; nMapIdx < nStrLen; nMapIdx++) {
    nChr = sDOMStr.codePointAt(nMapIdx);

    if (nChr > 65536) {
      nMapIdx++;
    }

    nArrLen += nChr < 0x80 ? 1 : nChr < 0x800 ? 2 : nChr < 0x10000 ? 3 : nChr < 0x200000 ? 4 : nChr < 0x4000000 ? 5 : 6;
  }

  aBytes = new Uint8Array(nArrLen);

  /* transcription… */
  let nIdx = 0;
  let nChrIdx = 0;
  while (nIdx < nArrLen) {
    nChr = sDOMStr.codePointAt(nChrIdx);
    if (nChr < 128) {
      /* one byte */
      aBytes[nIdx++] = nChr;
    } else if (nChr < 0x800) {
      /* two bytes */
      aBytes[nIdx++] = 192 + (nChr >>> 6);
      aBytes[nIdx++] = 128 + (nChr & 63);
    } else if (nChr < 0x10000) {
      /* three bytes */
      aBytes[nIdx++] = 224 + (nChr >>> 12);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
    } else if (nChr < 0x200000) {
      /* four bytes */
      aBytes[nIdx++] = 240 + (nChr >>> 18);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    } else if (nChr < 0x4000000) {
      /* five bytes */
      aBytes[nIdx++] = 248 + (nChr >>> 24);
      aBytes[nIdx++] = 128 + (nChr >>> 18 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    } else /* if (nChr <= 0x7fffffff) */ {
      /* six bytes */
      aBytes[nIdx++] = 252 + (nChr >>> 30);
      aBytes[nIdx++] = 128 + (nChr >>> 24 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 18 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    }
    nChrIdx++;
  }

  return aBytes;
}

const encodeJs = (input: string) => {
  return base64EncArr(strToUTF8Arr(input));
};

const decodeJs = (input: string) => {
  return utf8ArrToStr(base64DecToArr(input));
};

/**
 * encode string to base64
 *
 * @param input
 */
const encode = function (input = ""): string {
  if (globalThis.Buffer !== undefined) {
    return globalThis.Buffer.from(input, "utf-8").toString("base64");
  }
  return base64EncArr(strToUTF8Arr(input));
};

/**
 * decode base64 to string
 *
 * @param input base64 string
 */
const decode = function (input: string): string {
  if (globalThis.Buffer !== undefined) {
    return globalThis.Buffer.from(input, "base64").toString("utf-8");
  }
  return utf8ArrToStr(base64DecToArr(input));
};

export { encode, decode, encodeJs, decodeJs };
