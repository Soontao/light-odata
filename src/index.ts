import { OData } from "./odata";

export * from "./batch";
export * from "./filter";
export * from "./params";
export * from "./types";
export { EdmV4 } from "./types_v4";
export * from "./util";
export * from "./request";
export { transformation } from "./tranformation";
export { OData };

export default OData;
