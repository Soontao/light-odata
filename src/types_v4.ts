import type { BatchRequestV4, ParsedResponseV4 } from "./batch";
import type { SystemQueryOptions } from "./params";
import { OData } from "./odata";
import type {
  BatchRequestOptions, ODataActionImportRequest,
  ODataActionRequest, ODataFunctionImportRequest, ODataFunctionRequest,
  ODataQueryRequest, ODataReadIDRequest,
  ODataWriteRequest,
  UnwrapBatchRequest, UnwrapPromise
} from "./types";
import { EdmString, Guid, ODataValueObject, RawString } from "./types";


export interface PlainODataResponseV4 {

  error?: { /** if error occured, node error will have value */
    code: string;
    message: {
      lang: string,
      value: string /** server error message */
    } & string
  }

  /**
   * context string
   */
  "@odata.context"?: string;

  /**
   * total count
   */
  "@odata.count"?: number;

  "@odata.id"?: string;

  "@odata.etag"?: string;

  "@odata.editLink"?: string;

}

export type BatchPlainODataResponseV4<E = any> = PlainODataResponseV4 & {

  value?: Array<E>

} & E

export type PlainODataSingleResponseV4<E = any> = E & PlainODataResponseV4


export type PlainODataMultiResponseV4<E = any> = PlainODataResponseV4 & {

  /**
   * values result list
   */
  value?: Array<E>;

}

export interface ODataV4 extends OData {

  /**
   * get odata version
   */
  getVersion(): "v4";

  /**
   * create new odata http request
   */
  newRequest<T>(options: ODataQueryRequest): Promise<PlainODataMultiResponseV4<T>>;
  /**
   * create new odata http request
   */
  newRequest<T>(options: ODataWriteRequest<T>): Promise<PlainODataSingleResponseV4<T>>;
  /**
   * create new odata http request
   */
  newRequest<T>(options: ODataReadIDRequest): Promise<PlainODataSingleResponseV4<T>>;
  /**
   * create new odata http request
   */
  newRequest(options: ODataFunctionRequest): Promise<PlainODataResponseV4>;
  /**
   * create new odata http request
   */
  newRequest(options: ODataActionRequest): Promise<PlainODataResponseV4>;
  /**
   * create new odata http request
   */
  newRequest(options: ODataFunctionImportRequest): Promise<PlainODataResponseV4>;
  /**
   * create new odata http request
   */
  newRequest(options: ODataActionImportRequest): Promise<PlainODataResponseV4>;

  functionImport(functionName: string, parameters?: any, params?: SystemQueryOptions): Promise<PlainODataResponseV4>;
  actionImport(actionName: string, parameters?: any, params?: SystemQueryOptions): Promise<PlainODataResponseV4>;

  /**
   * create batch request (will not perform)
   *
   * @param options
   */
  newBatchRequest<T>(options: BatchRequestOptionsV4<T>): Promise<BatchRequestV4>;

  /**
   * execute batch requests and get response
   *
   * @param requests batch request
   *
   */
  execBatchRequests<T extends BatchRequestsV4 = any>(requests: T): BatchResponsesV4<T>


  /**
   * execute batch requests in OData `V4.01` JSON format, and get the response
   *
   * @param requests
   */
  execBatchRequestsJson<T extends BatchRequestsV4 = any>(requests: T): BatchResponsesV4<T>

}

export type BatchRequestsV4 = Array<Promise<BatchRequestV4>>

export type BatchResponsesV4<T> = Promise<{ [K in keyof T]: ParsedResponseV4<UnwrapBatchRequest<UnwrapPromise<T[K]>>> }>

export interface BatchRequestOptionsV4<T> extends BatchRequestOptions<T> {
  /**
  * odata-json-format-v4.01
  * request id
  */
  requestId?: string;
  /**
   * odata-json-format-v4.01
   * atomicityGroup
   */
  atomicityGroup?: string;
  /**
   * odata-json-format-v4.01
   * inner headers
   */
  headers?: Record<string, string>;
  /**
   * odata-json-format-v4.01
   * dependsOn
   */
  dependsOn?: Array<string>;
}

/**
 * date object for odata v4
 *
 * @version 4.0
 */
export abstract class ODataDateBase extends ODataValueObject {

  protected constructor(date: Date | string | number) {
    if (date instanceof Date) {
      super(date);
    }
    else {
      super(new Date(date));
    }
  }

}

/**
 * Edm.DateTime
 *
 * @version 4.0
 */
export class ODataDateTimeOffset extends ODataDateBase {

  public static from(date: Date): ODataDateTimeOffset {
    return new ODataDateTimeOffset(date);
  }

  public toString(): string {
    return this.rawValue.toISOString();
  }

}

/**
 * @version 4.0
 */
export class ODataDate extends ODataDateBase {

  public static from(date: Date): ODataDate {
    return new ODataDate(date);
  }

  toString(): string {
    return this.rawValue.toISOString().split("T")[0];
  }

}


export const EdmV4 = {
  RawString,
  DateTimeOffset: ODataDateTimeOffset,
  Date: ODataDate,
  Guid,
  String: EdmString,
};
