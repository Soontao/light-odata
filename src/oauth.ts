import { Mutex } from "@newdash/newdash/functional/Mutex";
import { encode } from "./base64";
import { AuthenticationError } from "./errors";
import { SearchParams } from "./util";

const S_AUTHORIZATION = "Authorization";
const S_BEARER = "Bearer";
const S_CLIENT_CREDENTIALS = "client_credentials";
const S_CT_URL_FORM = "application/x-www-form-urlencoded";

export type TokenRetrieveType = "header" | "form";

/**
 * Client Credential OAuth Client
 *
 * Will automatically process the token refresh logic
 *
 * refresh policy: half expire-in
 *
 * @example
 * ```js
 * // please remember share client instance for a same OAuth server
 * const client = new ClientCredentialsOAuthClient(
 *  "http://example/oauth/token",
 *  "client_id",
 *  "client_secret"
 * )
 * const response = await client.fetch("http://protected.by.token/resource", {
 *  method: "GET",
 * })
 * ```
 *
 *
 */
class ClientCredentialsOAuthClient {

  private readonly tokenUrl: string;

  private readonly clientId: string;

  private readonly clientSecret: string;

  private readonly scope: string;

  private readonly mut = new Mutex();

  private token: string;

  private expiresAt = 0;

  private tokenRetrieveType: TokenRetrieveType = "header";


  /**
   * a very simple oauth client credential client
   *
   * @param tokenUrl oauth token url
   * @param clientId oauth client id
   * @param clientSecret oauth client secret
   * @param retrieveType the clientId and clientSecret is put into header or form body
   * @param scope oauth scope
   */
  constructor(
    tokenUrl: string,
    clientId: string,
    clientSecret: string,
    retrieveType: TokenRetrieveType = "header",
    scope?: string,
  ) {

    this.tokenUrl = tokenUrl;
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.tokenRetrieveType = retrieveType;
    this.scope = scope;
  }

  /**
   * fetch token directly
   * @private
   * @returns {Promise<any>}
   */
  private async fetchOAuthResponse(): Promise<any> {
    const params = new SearchParams();

    params.append("grant_type", S_CLIENT_CREDENTIALS);

    if (typeof this.scope === "string") { params.append("scope", this.scope); }

    let response = undefined;

    if (this.tokenRetrieveType === "form") {

      params.append("client_id", this.clientId);
      params.append("client_secret", this.clientSecret);

      response = await fetch(this.tokenUrl, {
        method: "POST",
        body: params.toString(),
        headers: { "Content-Type": S_CT_URL_FORM }
      });

    }
    else {

      // standard way described in rfc6749
      // https://datatracker.ietf.org/doc/html/rfc6749#section-4.4.2
      response = await fetch(this.tokenUrl, {
        method: "POST",
        body: params.toString(),
        headers: {
          [S_AUTHORIZATION]: "Basic " + encode(`${this.clientId}:${this.clientSecret}`),
          "Content-Type": S_CT_URL_FORM
        }
      });

    }


    if (response.status >= 400) {

      if (response.headers.get("Content-Type")?.includes?.("application/json")) {
        const responseBody = await response.json();
        if ("error" in responseBody) {
          throw new AuthenticationError(responseBody.error);
        }
        else {
          throw new AuthenticationError(JSON.stringify(responseBody));
        }
      }
      else {
        throw new Error(await response.text());
      }
    }

    return response.json();
  }

  /**
   * get header pair
   */
  public async getHeader() {
    const token = await this.getToken();
    return {
      [S_AUTHORIZATION]: `${S_BEARER} ${token}`
    };
  }

  /**
   * get current valid oauth token
   *
   * @returns {Promise<string>}
   */
  public async getToken(): Promise<string> {

    // lock avoid multi fetch token requests in same time
    return this.mut.use(async () => {

      try {

        const currentTime = new Date().getTime();

        // retrive new token
        if (currentTime >= this.expiresAt) {
          const { access_token: accessToken, expires_in: expiresIn } = await this.fetchOAuthResponse();
          this.token = accessToken;
          // half expire
          this.expiresAt = currentTime + ((expiresIn / 2) * 1000);
        }

      } catch (error) {

        console?.error?.(`oauth client retrieve token failed: ${error?.message ?? error}`,);

        // rethrow error
        throw error;

      }

      return this.token;

    });


  }

  /**
   * send requests with oauth token
   *
   */
  public async fetch(...args: Parameters<typeof fetch>) {
    // refresh token if required
    const token = await this.getToken();

    if (token === undefined) {
      throw new TypeError("un-expected token is undefined");
    }

    const url = args[0];
    let config = args[1];

    if (config === undefined) {
      config = {};
    }
    // append jwt token
    config.headers = {
      ...(config.headers || {}),
      [S_AUTHORIZATION]: `${S_BEARER} ${token}`
    };

    return fetch(url, config);
  }

}

export { ClientCredentialsOAuthClient };
