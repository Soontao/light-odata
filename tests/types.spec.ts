import { formatId } from "../src";
import * as types from "../src/types";
import * as typesV4 from "../src/types_v4";

describe('Types Test Suite', () => {

  it('should support support Edm object', () => {
    expect(types.EdmV2).not.toBeUndefined()
    expect(typesV4.EdmV4).not.toBeUndefined()
  });

  it('should support convert paylaod object with value', () => {
    expect(JSON.stringify({
      v: 1,
      c: null,
      d: { v: "value" }
    }, types.ODataValueJSONReplacer))
      .toMatchSnapshot()

    expect(JSON.stringify({
      v: types.EdmV2.RawString.from('2'),
      c: types.EdmV2.DateTime.from(new Date("2022-02-25T00:00:00Z")),
      d: { v: "value" }
    }, types.ODataValueJSONReplacer))
      .toMatchSnapshot()
  });

  it('should support deep transform', () => {
    expect(JSON.stringify({
      v: types.EdmV2.RawString.from('2'),
      c: null,
      d: {
        v: typesV4.EdmV4.Date.from(new Date("2022-02-25T00:00:00Z")),
      }
    }, types.ODataValueJSONReplacer))
      .toMatchSnapshot()
  });

  it('should support EdmV2.DateTimeOffset', async () => {
    expect(JSON.stringify({
      v: types.EdmV2.DateTimeOffset.from(new Date("2022-02-25T00:00:00Z")),
    }, types.ODataValueJSONReplacer))
      .toMatchSnapshot()
  });

  it('should support EdmV4.DateTimeOffset', async () => {
    expect(JSON.stringify({
      v: typesV4.EdmV4.DateTimeOffset.from(new Date("2022-02-25T00:00:00Z")),
    }, types.ODataValueJSONReplacer))
      .toMatchSnapshot()
  });

  it('should support format Edm.String', () => {
    expect(JSON.stringify({ ID: types.EdmV2.String.from("v1") }, types.ODataValueJSONReplacer)).toMatchSnapshot()
    expect(JSON.stringify({ ID: typesV4.EdmV4.String.from("v1") }, types.ODataValueJSONReplacer)).toMatchSnapshot()
  });

  it('should support format-id for url', () => {
    expect(formatId(1)).
      toMatchInlineSnapshot(`"(1)"`)
    expect(formatId({ ID: 1 })).
      toMatchInlineSnapshot(`"(ID=1)"`)
    expect(formatId('1')).
      toMatchInlineSnapshot(`"('1')"`)
    expect(formatId({ ID: '1' })).
      toMatchInlineSnapshot(`"(ID='1')"`)
    expect(formatId({ ID: 'f0b3e560-b0c9-47ba-b90b-7913181e131f' })).
      toMatchInlineSnapshot(`"(ID=f0b3e560-b0c9-47ba-b90b-7913181e131f)"`)
    expect(formatId({ ID: types.Guid.from('f0b3e560-b0c9-47ba-b90b-7913181e131f') })).
      toMatchInlineSnapshot(`"(ID=f0b3e560-b0c9-47ba-b90b-7913181e131f)"`)
  });

});
