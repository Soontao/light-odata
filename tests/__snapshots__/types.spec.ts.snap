// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Types Test Suite should support EdmV2.DateTimeOffset 1`] = `"{\\"v\\":\\"/Date(1645747200000)/\\"}"`;

exports[`Types Test Suite should support EdmV4.DateTimeOffset 1`] = `"{\\"v\\":\\"2022-02-25T00:00:00.000Z\\"}"`;

exports[`Types Test Suite should support convert paylaod object with value 1`] = `"{\\"v\\":1,\\"c\\":null,\\"d\\":{\\"v\\":\\"value\\"}}"`;

exports[`Types Test Suite should support convert paylaod object with value 2`] = `"{\\"v\\":\\"2\\",\\"c\\":\\"/Date(1645747200000)/\\",\\"d\\":{\\"v\\":\\"value\\"}}"`;

exports[`Types Test Suite should support deep transform 1`] = `"{\\"v\\":\\"2\\",\\"c\\":null,\\"d\\":{\\"v\\":\\"2022-02-25\\"}}"`;

exports[`Types Test Suite should support format Edm.String 1`] = `"{\\"ID\\":\\"v1\\"}"`;

exports[`Types Test Suite should support format Edm.String 2`] = `"{\\"ID\\":\\"v1\\"}"`;
